cpp/src/join/distinct_hash_join.cu

```diff
@@ -170,23 +232,39 @@ distinct_hash_join<HasNested>::inner_join(rmm::cuda_stream_view stream,
   auto probe_indices =
     std::make_unique<rmm::device_uvector<size_type>>(probe_table_num_rows, stream, mr);
 
+  std::unordered_set<cudf::type_id> probe_column_types;
+  for (auto col : this->_probe) {
+    probe_column_types.insert(col.type().id());
+  }
+
   auto const probe_row_hasher =
     cudf::experimental::row::hash::row_hasher{this->_preprocessed_probe};
-  auto const d_probe_hasher = probe_row_hasher.device_hasher(nullate::DYNAMIC{this->_has_nulls});
-  auto const iter           = cudf::detail::make_counting_transform_iterator(
-    0, build_keys_fn<decltype(d_probe_hasher), lhs_index_type>{d_probe_hasher});
+  auto const d_probe_hasher =
+    probe_row_hasher.device_hasher(probe_column_types, nullate::DYNAMIC{this->_has_nulls});
 
   auto const build_indices_begin =
     thrust::make_transform_output_iterator(build_indices->begin(), output_fn{});
   auto const probe_indices_begin =
     thrust::make_transform_output_iterator(probe_indices->begin(), output_fn{});
 
-  auto const [probe_indices_end, _] = this->_hash_table.retrieve(
-    iter, iter + probe_table_num_rows, probe_indices_begin, build_indices_begin, {stream.value()});
-
-  auto const actual_size = std::distance(probe_indices_begin, probe_indices_end);
-  build_indices->resize(actual_size, stream);
-  probe_indices->resize(actual_size, stream);
+  std::visit(
+    [&](auto&& hasher, auto&& hash_table) {
+      auto const iter = cudf::detail::make_counting_transform_iterator(
+        0,
+        build_keys_fn<typename std::remove_reference_t<decltype(hasher)>, lhs_index_type>{hasher});
+
+      auto const [probe_indices_end, _] = hash_table.retrieve(iter,
+                                                              iter + probe_table_num_rows,
+                                                              probe_indices_begin,
+                                                              build_indices_begin,
+                                                              {stream.value()});
+
+      auto const actual_size = std::distance(probe_indices_begin, probe_indices_end);
```

Comment:
If these are device iterators, we probably want `thrust::distance`.

Reply:
I understand this, but why should we use `thrust::distance`, if pointer arithmetic should work fine using `std::distance` too?

Reply:
It future-proofs the code. Always use thrust::distance with device iterators, whether raw pointers or fancy iterators.